crates/subspace-farmer/src/utils/segment_reconstruction.rs

```diff
@@ -0,0 +1,79 @@
+use std::iter;
+use futures::stream::{FuturesOrdered, FuturesUnordered};
+use futures::StreamExt;
+use subspace_core_primitives::{Piece, PieceIndex, SegmentIndex, PIECES_IN_SEGMENT, RECORD_SIZE};
+use subspace_farmer_components::plotting::PieceGetter;
+use tracing::{error, info};
+use subspace_archiving::reconstructor::Reconstructor;
+
+const SEGMENT_SIZE: u32 = RECORD_SIZE * PIECES_IN_SEGMENT / 2;
+
+pub struct SegmentReconstruction<PG> {
+    piece_getter: PG,
+}
+
+impl<PG: PieceGetter> SegmentReconstruction<PG> {
+    pub fn new(piece_getter: PG) -> Self {
+        Self { piece_getter }
+    }
+
+    pub async fn get_missing_piece(&self, missing_piece_index: PieceIndex) -> Option<Piece> {
+        let segment_index: SegmentIndex =
+            missing_piece_index / SegmentIndex::from(PIECES_IN_SEGMENT);
+
+        // TODO: Consider taking more pieces here.
+        // We take minimum pieces to reconstruct the segment plus one to mitigate the missing piece.
+        let pieces_to_retrieve = (PIECES_IN_SEGMENT / 2 + 1) as usize;
+
+        let starting_piece_index = segment_index * SegmentIndex::from(PIECES_IN_SEGMENT);
+
+        let piece_offset = (missing_piece_index - starting_piece_index) as usize; // TODO:
+        info!(%missing_piece_index, %starting_piece_index, %pieces_to_retrieve, %piece_offset);
```

Comment:
Small nit. For some types you can skip `?` or `%` as they can be passed by value to `tracing` logs (`tracing` has some [built-in types](https://docs.rs/tracing/0.1.37/tracing/field/trait.Visit.html) and `std::fmt::{Debug, Display}` are only 2 of them)